echo-core/src/test/java/com/netflix/spinnaker/echo/services/IgorServiceTest.java

```diff
@@ -0,0 +1,117 @@
+import static com.github.tomakehurst.wiremock.client.WireMock.urlEqualTo;
+import static com.github.tomakehurst.wiremock.client.WireMock.verify;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import com.github.tomakehurst.wiremock.WireMockServer;
+import com.github.tomakehurst.wiremock.client.WireMock;
+import com.github.tomakehurst.wiremock.core.WireMockConfiguration;
+import com.netflix.spinnaker.echo.util.RetrofitUtils;
+import com.netflix.spinnaker.kork.artifacts.model.Artifact;
+import com.netflix.spinnaker.kork.retrofit.ErrorHandlingExecutorCallAdapterFactory;
+import com.netflix.spinnaker.kork.retrofit.Retrofit2SyncCall;
+import java.util.List;
+import java.util.Map;
+import okhttp3.OkHttpClient;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import retrofit2.Call;
+import retrofit2.Retrofit;
+import retrofit2.converter.jackson.JacksonConverterFactory;
+import retrofit2.http.GET;
+import retrofit2.http.Path;
+import retrofit2.http.Query;
+
+public class IgorServiceTest {
+  WireMockServer wireMockServer;
+  int port;
+  IgorService igorService;
+  String baseUrl = "http://localhost:PORT";
+
+  @BeforeEach
+  void setUp() {
+    wireMockServer = new WireMockServer(WireMockConfiguration.options().dynamicPort());
+    wireMockServer.start();
+    port = wireMockServer.port();
+    WireMock.configureFor("localhost", port);
+
+    baseUrl = baseUrl.replaceFirst("PORT", String.valueOf(port));
+
+    igorService =
+        new Retrofit.Builder()
+            .baseUrl(RetrofitUtils.getBaseUrl(baseUrl))
+            .client(new OkHttpClient())
+            .addCallAdapterFactory(ErrorHandlingExecutorCallAdapterFactory.getInstance())
+            .addConverterFactory(JacksonConverterFactory.create())
+            .build()
+            .create(IgorService.class);
+  }
+
+  @AfterEach
+  void tearDown() {
+    wireMockServer.stop();
+  }
+
+  @Test
+  void testIgorService_withQueryParamBeforePathVar() {
+    stubFor(
+        get(urlEqualTo("/builds/artifacts/1/propertyFile/master/job"))
+            .willReturn(aResponse().withStatus(200)));
+
+    assertThrows(
+        java.lang.IllegalArgumentException.class,
+        () ->
+            Retrofit2SyncCall.execute(igorService.getArtifacts(1, "propertyFile", "master", "job")),
+        "A @Path parameter must not come after a @Query");
+  }
+
+  @Test
+  void testIgorService_withQueryParamAfterAllPathVars() {
+    stubFor(
+        get(urlEqualTo("/builds/properties/1/fileName/master?job=job"))
+            .willReturn(aResponse().withStatus(200).withBody("{\"mesage\":\"success\"}")));
+
+    Retrofit2SyncCall.execute(
+        igorService.getPropertyFileWithJobQueryParameter(1, "fileName", "master", "job"));
+
+    verify(1, getRequestedFor(urlEqualTo("/builds/properties/1/fileName/master?job=job")));
+  }
+
+  private interface IgorService {
```

Comment:
Can't we use the "real" IgorService in a test?

Reply:
That's OK.  I think it's better to have tests that cover the actual code.

Reply:
Ok, will change shortly.